[sample consensus] Better performance, error handling and other improvements

mvieth opened the pull request.

kunaltyagi commented:
Might be better to not modify the public API (or add overloads instead of replacing the original functions)

mvieth commented:
I could add adapters for `getSamples`, `computeModelCoefficients`, `isSampleGood`, and the other functions, but I would mark them as deprecated right away.

kunaltyagi commented:
Might not, because based on feedback on RFC-002, we might make `pcl::index_t` be either 32 bit or 64 bit at build time. The deprecation will need to be undone in case someone is building `pcl::index_t` in 32-bit mode (which might be the default)

mvieth commented:
I wasn't aware that there are RFCs in the wiki. I thought this kind of stuff is discussed in issues. I made a comment in the issue linked in the RFC.

kunaltyagi commented:
> I thought this kind of stuff is discussed in issues

It is. I created RFC since it turns out a fleshing out a topic is hard to do in an issue. RFC allows people to comment and discuss in the relevant gitter rooms without creating noise for other non-interested people.

mvieth commented:
Well, nobody is forced to read an issue or its comments. As long as an issue is confined to one topic, I see no problem with discussing things that way. I just wanted to note that many people may not know about the RFCs in the wiki and thus don't participate in the discussion.

mvieth commented:
> Is there a reason behind changing the behavior of isModelValid?

What do you mean? Where did I do that?

kunaltyagi commented:
You deleted the `pcl::SampleConsensusModelNormalSphere<PointT, PointNT>::isModelValid` function which effectively changes the behavior of `isModelValid` since it defaults to the base-class function

mvieth commented:
I added a `using SampleConsensusModelSphere<PointT>::isModelValid;` in sac_model_normal_sphere.h, and that one looks exactly like the function I deleted. So the behaviour should be the same, no?

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_cone.hpp:
```diff
@@ -45,18 +45,23 @@
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointT, typename PointNT> bool
-pcl::SampleConsensusModelCone<PointT, PointNT>::isSampleGood(const std::vector<int> &) const
+pcl::SampleConsensusModelCone<PointT, PointNT>::isSampleGood(const std::vector<std::size_t> &samples) const
```
Ok, I removed that commit.

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle3d.hpp:
```diff
@@ -169,7 +174,7 @@ pcl::SampleConsensusModelCircle3D<PointT>::selectWithinDistance (
     return;
   }
-  int nr_p = 0;
+  std::size_t nr_p = 0;
```
Done.

kunaltyagi commented:
PS: I started work on a `pcl::index_t` and I'll create a PR soon. May take a day or two. Could you wait for that PR to merge so you can use that instead of breaking the public API.

kunaltyagi commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp:
```diff
@@ -156,13 +162,10 @@ pcl::SampleConsensusModelCircle2D<PointT>::selectWithinDistance (
-      inliers[nr_p] = (*indices_)[i];
-      error_sqr_dists_[nr_p] = static_cast<double> (distance);
-      ++nr_p;
+      inliers.push_back ((*indices_)[i]);
```
Some loops can be modified to remove the loop variable `i` and replaced by a range-based for loop. Not all though. Might be something to add

mvieth replied:
Yes, that would be possible. But I think for this PR I will leave it like it is.

kunaltyagi commented on sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp:
```diff
@@ -83,7 +84,7 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
     }
 
-    double d_cur_penalty = 0;
+    double d_cur_penalty;
```
I have mixed feelings about this. One one hand, we are declaring without initialization, on another hand, this allows linters to complain about uninitialized value if a branch downstream is not taken.

mvieth replied:
I agree that direct initialization is usually good, but in this case I really think that no initialization is the right thing to do because that value is never read, so it might even be confusing and misleading. Zero would anyway be the wrong value because it means no penalty=perfect model.

SergioRAgostinho replied:
Replace the block line 111-114 with a ternary operator and construct the value directly there.

kunaltyagi commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp:
```diff
@@ -137,9 +142,10 @@ pcl::SampleConsensusModelCircle2D<PointT>::selectWithinDistance (
-  int nr_p = 0;
-  inliers.resize (indices_->size ());
-  error_sqr_dists_.resize (indices_->size ());
+  inliers.clear ();
```
GCC has a weird bug/feature where `reserve` results in bypassing the exponential growth of the container and the exact memory is granted. This causes quadratic time if the function with `reserve` is called in a loop.

While I do not envision this being called in a loop with growing `indices_.size()`, is it possible to either
* put a warning note somewhere
* not use reserve and pay for multiple allocations in the loop (just `push_back`, nothing else)
* (current) pay for zero constructing the array using `resize`

None of the 3 options are "good". I'd really like C++ STL to give a mix of resize + reserve: don't initialize, just gimme exponentially increasing memory.

mvieth replied:
Interesting. Do you have a link to an article regarding that behaviour or something? Never heard of it before. Is this only an issue together with `clear()`?

kunaltyagi replied:
No, this is an issue with `reserve`. It's there in the docs: [reserve](https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/vector.tcc#L78) allocates exact memory, but doubles capacity for other cases.

mvieth replied:
Sorry, I am still not sure what the problem is. We want that exact amount of memory, no? And `reserve` would only do something if the new desired capacity is bigger than the current capacity. So even if `indices_->size()` gets bigger (and I agree with you, I don't see how that would happen), this would work. And where does this exponential growth happen?

kunaltyagi replied:
On using `reserve` in a loop, the loop can take quadratic time instead of amortized linear since growth of the vector becomes quantized linear (step of one in worst case) instead of exponential.

We do want to allocate exact memory. The hypothetical quadratic time issue happens **IFF** `indices_->size ()` grows linearly in the loop. I can't comprehend a situation where this can happen. If @taketwo or @SergioRAgostinho can't think of a valid case where the bad-case happens, we can go ahead with this patch.

My position is of safety since I couldn't be certain that the hypothetical can never happen. To clarify:
* I asked if it was possible to replace `reserve` with something else based on a hypothetical. **If that never happens, this is resolved.**
* The issue isn't with `clear` but only `reserve` in case the hypothetical use happens in user-code. See implementation for GCC's weird behavior; clang is saner.

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp:
```diff
@@ -96,22 +97,36 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
     // d_cur_penalty = median (distances)
-    std::size_t mid = sac_model_->getIndices ()->size () / 2;
-    if (mid >= distances.size ())
+    const std::size_t mid = nr_valid_dists / 2;
```
I would say it makes more sense to take the middle of only the valid distances, omitting the NaNs. But I guess `mid >= distances.size ()` is unnecessary, since `nr_valid_dists` should always be smaller or equal to `distances.size()`, so testing `nr_valid_dists == 0` alone should be enough, right?

kunaltyagi replied:
The issue was the old impl switching from index to pointcloud as if they are one and the same. See old 111, 112. But you are correct. Testing `nr_valid_dists` seems sufficient

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp:
```diff
@@ -96,22 +97,36 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
       continue;
     }
+    // Move all NaNs in distances to the end
+    const auto new_end = std::partition (distances.begin(), distances.end(), [](double d){return !std::isnan (d);});
```
Makes sense. Done.

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp:
```diff
@@ -96,22 +97,36 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
+      // Looking at two values instead of one probably doesn't matter because they are mostly barely different, but let's do it for accuracy's sake
+      std::nth_element (distances.begin (), distances.begin () + (mid - 1), new_end);
+      const double tmp = distances[mid-1];
+      std::nth_element (distances.begin (), distances.begin () + mid, new_end);
```
This would be possible: `partial_sort(distance.begin(), distances.begin()+mid, new_end)`, but using `nth_element` twice seems to be faster.

kunaltyagi replied:
My mistake for not being clear. I'd recommend you to change the indices used for your second `nth_element`. Currently, it wastes effort by going through the first half, when it only needs to do one of the following:
* partial sort 1 element in the second half
* get 1st element in the second half

mvieth replied:
I guess `std::min_element(distances.begin() + mid, new_end)` would be the best solution then.

kunaltyagi replied:
Yeah, that'd be better since less moving/copying. You're 🔥

kunaltyagi commented on sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp:
```diff
@@ -96,22 +97,36 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
     }
+    // Move all NaNs in distances to the end
+    const auto new_end = (sac_model_->getInputCloud()->is_dense ? distances.end() : std::partition (distances.begin(), distances.end(), [](double d){return !std::isnan (d);}));
+    const auto nr_valid_dists = (sac_model_->getInputCloud()->is_dense ? distances.size() : std::distance (distances.begin (), new_end));
```
The check for `is_dense` here is not needed since `new_end` (I'd recommend naming it `number_end` or something) is already `distances.end()` for the computation.

mvieth replied:
Yes, it is not needed, but why call `std::distance` when `distances.size()` already has the desired value?

kunaltyagi replied:
`size` usually calls the difference since storing `front` and `back` pointers is often what the implementations do. See [godbolt](https://godbolt.org/z/pBU-Jm) to see that the codegen is identical. So we can opt for the simpler solution:

```suggestion
    const auto nr_valid_dists = std::distance (distances.begin (), new_end);
```

mvieth commented:
@kunaltyagi Ok, I think I applied all of your suggestions. But what is going on with the MacOS checks?

SergioRAgostinho commented:
> @kunaltyagi Ok, I think I applied all of your suggestions. But what is going on with the MacOS checks?

Broken build tools environment in CI.

kunaltyagi commented on sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp:
```diff
@@ -96,22 +97,36 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
     // Do we have a "middle" point or should we "estimate" one ?
-    if (sac_model_->getIndices ()->size () % 2 == 0)
-      d_cur_penalty = (sqrt (distances[mid-1]) + sqrt (distances[mid])) / 2;
+    if ((nr_valid_dists % 2) == 0)
```
`nr_valid_dists & 1` is better since modulo is a slow op and even clang trunk on O2 doesn't optimize it.

mvieth replied:
I searched around and found [this answer](https://stackoverflow.com/a/160935). But tbh: it is done only once every iteration, so a few hundred times max. I do not see how it would have any effect on the performance.

kunaltyagi replied:
> I do not see how it would have any effect on the performance.

It doesn't, it's just something that's a reflex to me. asm difference is [visible](https://godbolt.org/z/KWE3rQ) under some conditions, but not always. So using bitwise and everywhere brings in uniformity. Like I said, both of these are minor comments.

kunaltyagi commented on sample_consensus/include/pcl/sample_consensus/impl/lmeds.hpp:
```diff
@@ -96,22 +97,36 @@ pcl::LeastMedianSquares<PointT>::computeModel (int debug_verbosity_level)
+    if ((nr_valid_dists % 2) == 0)
+    {
+      // Looking at two values instead of one probably doesn't matter because they are mostly barely different, but let's do it for accuracy's sake
+      std::nth_element (distances.begin (), distances.begin () + (mid - 1), new_end);
```
Just a side: Would it make sense to move `nth_element` till `mid` out of the 2 branches? Then `tmp2` can be min_element from `begin` to `mid - 1`. This method separates the 2 cases so this is fine too.

mvieth replied:
I think it is good like it is now, because the separation might help understanding it.

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle3d.hpp:
```diff
@@ -257,14 +257,14 @@ pcl::SampleConsensusModelCircle3D<PointT>::optimizeModelCoefficients (
 
   // Need at least 3 samples
-  if (inliers.size () <= 3)
+  if (inliers.size () <= sample_size_)
```
There are also occurrences of this in circle, line, and plane. But in my opinion, `<=` makes more sense. Consider the comment in `sac_model_plane.hpp`: `// Need more than the minimum sample size to make a difference`. So if `inliers.size ()` is exactly `sample_size_`, that won't be a problem, but the model coefficient will not change, so it makes no sense to finish the function.
BTW, in cone and cylinder, it is tested if inliers is empty, so whatever operator we agree on (`<` or `<=`), I am going to change that there.

SergioRAgostinho replied:
Ok, so this is a refinement step given the original model. Requesting more than the minimum number of points makes sense.

I believe there's some argument to be made whether we should really throw an error or simply warn the user. And there's a second argument to be made on whether we should copy `model_coefficients` to `optimized_coefficients`, if the minimum number of points conditions is not verified.

mvieth replied:
Regarding the first point: if there are too few inliers, then the new model coefficients might actually be worse than the original ones. If there are `sample_size_` inliers, then throwing an error would not be strictly necessary, but in practice that will likely almost never happen, so checking for that is unnecessary IMO.
Regarding the second point: not copying the coefficients would make further checks necessary where `optimizeModelCoefficients` is called. With copying, the function basically works like "Here are some coefficients, please give me better ones or the same ones back, but never give me just zeros!".

SergioRAgostinho replied:
> Regarding the second point: not copying the coefficients would make further checks necessary where `optimizeModelCoefficients` is called. With copying, the function basically works like "Here are some coefficients, please give me better ones or the same ones back, but never give me just zeros!".

Let's copy them out then.

mvieth replied:
They are copied at the start of the function

mvieth commented on sample_consensus/include/pcl/sample_consensus/sac_model_normal_sphere.h:
```diff
@@ -157,13 +157,7 @@ namespace pcl
-      bool
-      isModelValid (const Eigen::VectorXf &model_coefficients) const override;
-
+      using SampleConsensusModelSphere<PointT>::isModelValid;
```
Without it, I get the following errors:
`error: there are no arguments to ‘isModelValid’ that depend on a template parameter, so a declaration of ‘isModelValid’ must be available [-fpermissive]`
`error: ‘isModelValid’ was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]`
`note: declarations in dependent base ‘pcl::SampleConsensusModelSphere<pcl::PointXYZ>’ are not found by unqualified lookup`
`note: use ‘this->isModelValid’ instead`
So `this->isModelValid` seems to be possible, but the line with `using` is nicer IMO.

SergioRAgostinho replied:
Agreed

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_stick.hpp:
```diff
@@ -106,6 +106,7 @@ pcl::SampleConsensusModelStick<PointT>::getDistancesToModel (
   // Needs a valid set of model coefficients
   if (!isModelValid (model_coefficients))
   {
+    PCL_DEBUG ("[pcl::SampleConsensusModelStick::getDistancesToModel] Given model is invalid!\n");
```
Done

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_cylinder.hpp:
```diff
@@ -452,9 +452,9 @@ pcl::SampleConsensusModelCylinder<PointT, PointNT>::isModelValid (const Eigen::V
     coeff[1] = model_coefficients[4];
     coeff[2] = model_coefficients[5];
-    coeff[3] = 0;
+    coeff[3] = 0.0f;
```
See my last commit, I have found an even simpler solution

SergioRAgostinho commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_stick.hpp:
```diff
@@ -230,15 +252,15 @@ pcl::SampleConsensusModelStick<PointT>::optimizeModelCoefficients (
+  // Need more than the minimum sample size to make a difference
+  if (inliers.size () <= sample_size_)
   {
     PCL_ERROR ("[pcl::SampleConsensusModelStick::optimizeModelCoefficients] Not enough inliers found to support a model (%lu)! Returning the same coefficients.\n", inliers.size ());
```
Here as well.

mvieth commented on sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp:
```diff
@@ -201,14 +204,14 @@ pcl::SampleConsensusModelCircle2D<PointT>::optimizeModelCoefficients (
+  // Need more than the minimum sample size to make a difference
+  if (inliers.size () <= sample_size_)
   {
     PCL_ERROR ("[pcl::SampleConsensusModelCircle2D::optimizeModelCoefficients] Not enough inliers found to support a model (%lu)! Returning the same coefficients.\n", inliers.size ());
```
Done

mvieth commented:
Done. Let me know if there are more commits you would like me to squash.
